💃 🥤 Extract interaction function from models

mberr opened the pull request:
![maxresdefault](https://user-images.githubusercontent.com/5069736/99291779-9ce1e500-2840-11eb-98ce-b6bea6f727d3.jpg)

This is a replacement for #88 , where the merge target is `master`.

> @mali-git As discussed in today's call, I tried to draft an API for the interaction function. It is built around the most generic form of interaction function, which has one batch dimension, and then allows broadcasting over multiple entities/relations to meet the use cases for e.g. scoring all tail entities at once, but also supports, e.g. full CWA scores.
>
> This can likely also help for the fast LCWA @lvermue once envisioned 😉
> 
> I did not define the methods to be static to allow for parametric interaction functions such as e.g. ER-MLP having some weights.

### Overview
* One shared implementation for `score_hrt` / `score_h` / `score_r` / `score_t` in the base class, done in `InteractionFunction`.
* One shared implementation of `_score` for all models sharing the same set of embeddings (e.g. TransE/DistMult/ERMLP -> one vector for each entity/relation, TransH -> additional vector for each entity, etc.)
* A state-less functional form of the interaction function where all necessary states are passed from the outside. This is done in `pykeen.nn.modules`.
* A state-full implementation of interaction function encapsulating all shared parameters (e.g. weight matrices for ERMLP / ConvE, etc.), but delegating the actual interaction to the state-less version. This is done in `pykeen.nn.functional`.

### Tasks: 
* [x] What to do with interaction models where we have more than one vector for an entity/relation, such as e.g. TransH?
* [x] Re-introduce slicing (In best case, on the generic level)
* [x] Re-introduce regularization on generic level
* [x] Fix R-GCN (or keep it broken for #110 ?)
* [x] Add regularizer/constrainer directly to modules, recursively search for modules having regularize and accumulate value.
* [x] Update to reshape in the generic `Interaction`
* ~[ ] Update pipeline model composition~ bumped to https://github.com/pykeen/pykeen/pull/163

### Dependencies:
* [x] #137

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -1139,3 +1140,169 @@ def _can_slice(fn) -> bool:
+        t = all_entities.unsqueeze(dim=0)
+
+        # get scores
+        scores = self(h=h, r=r, t=t)
```
is there a difference between self() and self.forward()? If not, I'd prefer self.forward() to keep things explicit

mberr replied:
Yes, cf. https://github.com/pytorch/pytorch/blob/495070b388478532f0149a37d75a38c7bd23ad96/torch/nn/modules/module.py#L165-L174

cthoyt replied:
Whooosh! Big difference

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -1174,3 +1178,168 @@ def _can_slice(fn) -> bool:
+
+
+class SimpleVectorEntityRelationEmbeddingModel(EntityRelationEmbeddingModel):
+    """A base class for embedding models which store a single vector for each entity and relation."""
```
single or simple?

cthoyt commented on src/pykeen/models/unimodal/complex.py:
```diff
@@ -7,18 +7,44 @@
 
 
-class ComplEx(EntityRelationEmbeddingModel):
+class ComplexInteractionFunction(InteractionFunction):
```
should there be a different base interaction function for complex embeddings?

mberr replied:
to verify that we only apply interaction functions which regard the vector as complex number to complex numbers?

mberr replied:
I am actually curious to see whether we can switch to "real" complex vectors https://pytorch.org/docs/stable/generated/torch.complex.html, which would handle the type check / promotion for us.

cthoyt replied:
We can continue discussion on https://github.com/pykeen/pykeen/pull/134. As of now, this won't be possible because auto-differentiation on the `torch.nn.functional.embedding` is not possible for complex tensors.

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -1195,3 +1197,174 @@ def _can_slice(fn) -> bool:
+        """
+        raise NotImplementedError
+
+    def score_t(
```
why only `score_t` and not `score_r` and `score_h`?

mberr replied:
It should also have score_r / score_h, but I wanted to await feedback on score_t before doing the other two.

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -1195,3 +1197,174 @@ def _can_slice(fn) -> bool:
+        return self._score(h_ind=hr_batch[:, 0], r_ind=hr_batch[:, 1], t_ind=None).view(-1, self.num_entities)
+
+    def score_h(self, rt_batch: torch.LongTensor) -> torch.FloatTensor:  # noqa: D102
+        return self._score(h_ind=None, r_ind=rt_batch[:, 0], t_ind=rt_batch[:, 1]).view(-1, self.num_entities)
```
what about `score_r()`?

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -1195,3 +1197,174 @@ def _can_slice(fn) -> bool:
+
+    def _reset_parameters_(self):
+        super()._reset_parameters_()
+        self.interaction_function.reset_parameters()
```
will any interaction functions need a post_parameter_update function?

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -1195,3 +1197,174 @@ def _can_slice(fn) -> bool:
+
+
+class InteractionFunction(nn.Module):
+    """Base class for interaction functions."""
```
what do you think about the InteractionFunction having access to the model? This way, some of the parameters and embeddings can be used by the functions without having to do a lot of passing around

mberr replied:
Keeping it functional eases testing and results in a clearer data flow.

mberr replied:
Also it makes it easier to use pykeen interaction functions outside of pykeen, and to use interaction functions from the outside with pykeen.

cthoyt replied:
You’ve got a good point. However, I was trying to come up with a way to use this abstract for TransD, whose interaction function needs the embeddings and the indexes to make projections, and it wasn’t fitting so well

mberr replied:
Extend the interaction function interface to `(h, r, t, **kwargs)`?

cthoyt replied:
Maybe for this PR we should just focus on the simplest case, where the interaction function only needs (h,r,t) embeddings. Then in a follow-up PR, we can create a new abstraction that can handle more abstract operations

mberr replied:
Too late, I added ConvE :sweat_smile:

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -1195,3 +1199,492 @@ def _can_slice(fn) -> bool:
+
+    def _check_shape(
+        self,
+        *x: Union[str, torch.Tensor],
```
wouldn't it be more specific to type it like
`*x: Tuple[torch.Tensor, str]`

then you don't have to do this unpacking

cthoyt commented on src/pykeen/models/unimodal/ermlp.py:
```diff
@@ -2,24 +2,94 @@
+        h: torch.FloatTensor,
+        r: torch.FloatTensor,
+        t: torch.FloatTensor,
+        **kwargs,
```
can you just omit **kwargs here?

mberr replied:
At least my IDE then complains about not implementing the interface :sweat_smile:

mberr replied:
Maybe we can just solve this by an annotation `no_more_kwargs` which wraps the function with the generic **kwargs interface, and does the check for additional kwargs.

cthoyt replied:
yuck, maybe we should just make the base class h,r,t,*args,**kwargs so the subclasses can be anything

cthoyt commented on src/pykeen/models/unimodal/trans_e.py:
```diff
@@ -2,26 +2,25 @@
 ]
 
 
-class TransE(EntityRelationEmbeddingModel):
```
Nice I had the same thought!

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -344,7 +351,11 @@ def supports_subbatching(self) -> bool:  # noqa: D400, D401
     def _reset_parameters_(self):  # noqa: D401
         """Reset all parameters of the model in-place."""
-        raise NotImplementedError
+        for module in self.modules():
```
will have to add `super()._reset_parameters_()` to all of the base classes

cthoyt replied:
Done in f8c1fada

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -344,7 +351,11 @@ def supports_subbatching(self) -> bool:  # noqa: D400, D401
     def _reset_parameters_(self):  # noqa: D401
         """Reset all parameters of the model in-place."""
-        raise NotImplementedError
+        for module in self.modules():
```
@mberr is this recursive?

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -1195,3 +1237,183 @@ def _can_slice(fn) -> bool:
+        r_indices: Optional[torch.LongTensor],
+        t_indices: Optional[torch.LongTensor],
+    ) -> torch.FloatTensor:  # noqa: D102
+        return 0.5 * sum(
```
this looks a lot like an average....

why stop at 2 sided model, why not have an n-sided model?

mberr replied:
Because it is once with the triple, and once with the flipped triple. There is not natural third way of flipping :wink:

mberr replied:
It started from SimplE, which essentially is a two-sided model with DistMult base-model.

mali-git commented on src/pykeen/nn/modules.py:
```diff
@@ -0,0 +1,793 @@
+
+        :param h: shape: (batch_size, d_e)
+            The head representations.
+        :param all_relations: shape: (batch_size, d_r)
```
(num_relations, d_r)?

mberr replied:
35dfbbb

mali-git commented on src/pykeen/nn/modules.py:
```diff
@@ -0,0 +1,793 @@
+            The head representations.
+        :param all_relations: shape: (batch_size, d_r)
+            The relation representations.
+        :param t: shape: (num_entities, d_e)
```
(batch_size, d_e)?

mberr replied:
6fd0f5a

cthoyt commented on src/pykeen/models/unimodal/simple.py:
```diff
@@ -75,61 +75,29 @@ def __init__(
-
-        # Regularization
-        self.regularize_if_necessary(h, r, t)
+        scores = super().forward(
```
Can we register the clamp as a post-forward hook?

mberr replied:
Maybe, cf. https://pytorch.org/docs/stable/generated/torch.nn.Module.html?highlight=register_forward_hook#torch.nn.Module.register_forward_hook

cthoyt commented on src/pykeen/models/unimodal/trans_d.py:
```diff
@@ -126,103 +79,26 @@ def __init__(
+            relation_embedding_specification=EmbeddingSpecification(
+                initializer=xavier_normal_,
+                constrainer=clamp_norm,
+                constrainer_kwargs=dict(maxnorm=1., p=2, dim=-1),
```
the entity and relation projections should have different specifications

cthoyt commented on src/pykeen/models/unimodal/trans_h.py:
```diff
@@ -87,27 +88,18 @@ def __init__(
         self.scoring_fct_norm = scoring_fct_norm
 
         # embeddings
-        self.normal_vector_embeddings = Embedding.init_with_device(
```
seems like this change just made more code

mberr replied:
here, yes. But it makes the contructors much more readable (in particular for those with two entity and relation embeddings)

cthoyt replied:
okay that's a good point too

cthoyt commented on src/pykeen/models/unimodal/trans_h.py:
```diff
@@ -87,27 +88,18 @@ def __init__(
+        self.normal_vector_embeddings = Embedding.from_specification(
             num_embeddings=triples_factory.num_relations,
             embedding_dim=embedding_dim,
-            device=self.device,
```
why not pass device?

mberr replied:
I think it might be worth to get rid of passing around device everywhere, and move the model once to the device it is needed (e.g. at the start of the training loop).

mberr replied:
It increases fault tolerance, since pytorch will take care for us of moving every submodule / parameter / buffer to the device.

cthoyt replied:
Agreed

cthoyt commented on src/pykeen/nn/emb.py:
```diff
@@ -103,12 +123,51 @@ def __init__(
         )
 
+    @classmethod
+    def from_specification(
```
the instantiation of an embedding from the specification should be a member function of the EmbeddingSpecification class

mberr replied:
You mean something like `EmbeddingSpecification.create(num, dim)` ?

mberr replied:
done in 
0ada540  ?

cthoyt replied:
yup

mali-git commented on src/pykeen/utils.py:
```diff
@@ -430,3 +436,98 @@ def random_non_negative_int() -> int:
+            errors.append(f"Invalid number of dimensions: {tensor.shape} vs. {shape}")
+            continue
+        for dim, name in zip(tensor.shape, shape):
+            exp_dim = dims.get(name)
```
`dims` is always empty

mberr replied:
8ef901f

mberr commented on src/pykeen/models/base.py:
```diff
@@ -247,7 +255,8 @@ def __init__(
         entity_representations: Union[None, RepresentationModule, Sequence[RepresentationModule]] = None,
         relation_representations: Union[None, RepresentationModule, Sequence[RepresentationModule]] = None,
-        interaction_function: InteractionFunction[HeadRepresentation, RelationRepresentation, TailRepresentation] = None,
+        interaction_function: Interaction[
```
Change parameter name to `interaction` as well?

mberr commented on src/pykeen/models/base.py:
```diff
@@ -495,10 +496,10 @@ def predict_scores_all_tails(
         self.eval()
-        if slice_size is None:
-            scores = self.score_t(hr_batch)
+        if slice_size is not None and self.can_slice_t:
```
With the recent changes, every model should allow slicing for h/r/t.

mberr replied:
Also, `slice_size` is now part of the `score_*` API, at least for `Interaction`, so we shouldn't need these if-else parts here.

mberr commented on src/pykeen/triples/triples_factory.py:
```diff
@@ -283,9 +284,9 @@ def __init__(
-                    set(self.relation_to_inverse.keys()).union(set(self.relation_to_inverse.values())),
-                )
+                _keys = set(self.relation_to_inverse.keys()) if self.relation_to_inverse else set()
+                _values = set(self.relation_to_inverse.values()) if self.relation_to_inverse else set()
```
memory-wise, the new formulation is worse, since the variables `_keys`/`_values`  stay referenced until the method returns.

mberr commented on .travis.yml:
```diff
@@ -2,7 +2,7 @@ os: linux
 dist: xenial
 language: python
-python: "3.7"
+python: "3.8"
```
:heart_eyes:

cthoyt replied:
was necessary for protocols. also now we can use from future import annotations :)

mberr commented on src/pykeen/nn/modules.py:
```diff
@@ -43,17 +41,29 @@
+    return x if isinstance(x, Sequence) else (x,)
+
+
+def _ensure_tuple(*x: Union[Representation, Sequence[Representation]]) -> Sequence[Sequence[Representation]]:
```
@cthoyt I am not sure how, but maybe we can get rid of the packing/unpacking sometimes necessary.

cthoyt replied:
yeah that's what I'm hoping to move towards. You saw that I externalized the `_get_batches()` function, so this `_ensure_tuple()` function is only used once. We need unit testssss

mberr commented on src/pykeen/models/unimodal/trans_h.py:
```diff
@@ -80,89 +81,21 @@ def __init__(
-        super().regularize_if_necessary(
-            self.entity_embeddings(indices=None),
-            self.normal_vector_embeddings(indices=None),  # FIXME
-            self.relation_embeddings(indices=None),
```
example for regularization of all embeddings, irrespective whether they were used in the current batch

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -202,29 +217,42 @@ def _new_init(self, *args, **kwargs):
-    #: The default parameters for the default regularizer class
-    regularizer_default_kwargs: ClassVar[Optional[Mapping[str, Any]]] = None
-    #: The instance of the regularizer
-    regularizer: Regularizer
```
if it's not saved in the model, where should default regularization information be saved?

mali-git commented on src/pykeen/models/multimodal/complex_literal.py:
```diff
@@ -2,21 +2,61 @@
+        x, literal = x[..., :self.embedding_dim], x[..., self.embedding_dim:]
+        x_re, x_im = split_complex(x)
+        x_re = self.real(torch.cat([x_re, literal], dim=-1))
+        x_im = self.real(torch.cat([x_im, literal], dim=-1))
```
Replace `self.real()` with `self.imag()` in line 53

mberr replied:
5c2ecd9

mali-git commented on src/pykeen/nn/emb.py:
```diff
@@ -34,30 +41,99 @@ def forward(
         """
         raise NotImplementedError
 
+    def get_in_canonical_shape(
```
Add short explanation about canonical shape?

mberr replied:
31136b4

mali-git commented on src/pykeen/nn/emb.py:
```diff
@@ -34,30 +41,99 @@ def forward(
+
+    regularizer: Optional[Regularizer] = None
+
+    def make(
```
What is the advantage of allowing to create an Embedding directly from the Specification class, and not only from the Embedding class (which is more intiutive in my opinion)?

mali-git commented on src/pykeen/regularizers.py:
```diff
@@ -190,34 +197,33 @@ class TransHRegularizer(Regularizer):
         # The regularization in TransH enforces the defined soft constraints that should computed only for every batch.
         # Therefore, apply_only_once is always set to True.
-        super().__init__(device=device, weight=weight, apply_only_once=True)
+        super().__init__(weight=weight, apply_only_once=False, parameters=[])
```
Set `apply_only_once=True`

mberr replied:
5db61c4

cthoyt commented on src/pykeen/models/unimodal/trans_h.py:
```diff
@@ -80,89 +80,25 @@ def __init__(
-            self.normal_vector_embeddings(indices=None),  # FIXME
-            self.relation_embeddings(indices=None),
+        self.regularizer = self._instantiate_default_regularizer(
+            entity_embeddings=list(self.entity_representations[0].parameters()).pop(),
```
@mberr would you please document what's going on in: `list(self.entity_representations[0].parameters()).pop()`

mberr replied:
It takes the entity embeddings (=first entity representations). `parameters()` is only an iterator, so I wrapped it in `list(...).pop()` to access the one element contained in this iterator.

mberr replied:
I am very open towards replacing that by something nicer (list a `first` / `head` function, but I did not find such)

cthoyt replied:
So the list of parameters is only one long?

mberr replied:
Yes, since it is TransH, there should be exactly one entity representation (the entity embedding).

cthoyt commented on src/pykeen/nn/modules.py:
```diff
@@ -93,7 +93,34 @@ class Interaction(nn.Module, Generic[HeadRepresentation, RelationRepresentation,
+        assert all(torch.is_tensor(x) for x in (h, r, t))
+        return dict(h=h, r=r, t=t)
+
+    def _prepare_state_for_functional(self) -> MutableMapping[str, Any]:
```
leave this as the abstract function

cthoyt replied:
wait nevermind

cthoyt commented on src/pykeen/nn/functional.py:
```diff
@@ -675,8 +675,8 @@ def structured_embedding_interaction(
     return negative_norm_of_sum(
-        extended_einsum("brde,bhd->bhre", r_h, h).unsqueeze(dim=3),
-        -extended_einsum("brde,btd->brte", r_t, t).unsqueeze(dim=1),
+        extended_einsum("bred,bhd->bhre", r_h, h).unsqueeze(dim=3),
```
<img width="1291" alt="Screen Shot 2020-11-18 at 19 11 41" src="https://user-images.githubusercontent.com/5069736/99570220-e9a8f580-29d1-11eb-85f0-ec38bb2cfaea.png">

mberr replied:
:laughing:

cthoyt committed:
Update adding regularizers

@mberr the weight_regularizers instance was only used in tests, so I updated the TransH class to use it explicitly. I know that the regularizers are looked up automatically, but this makes it more obvious why the weight_regularizers variable is even there

cthoyt committed:
Update trans_h.py

cthoyt commented on src/pykeen/models/base.py:
```diff
@@ -1032,166 +1093,455 @@ def load_state(self, path: str) -> None:
+        # model.modules(). Thereby, we can collect them automatically.
+        self.weight_regularizers = nn.ModuleList()
+
+    def append_weight_regularizer(
```
@mberr this function is only used during tests. Can it be moved out of the base class?

mberr replied:
No, because it is the proper/public way to add weight regularizer's.

cthoyt commented on src/pykeen/models/unimodal/conv_kb.py:
```diff
@@ -81,66 +78,23 @@ def __init__(
-
-    def _reset_parameters_(self):  # noqa: D102
-        # embeddings
+        self.regularizer = self._instantiate_default_regularizer(parameters=list(self.interaction.parameters()))
```
Why isn't this stored in the weight_regularizers list?

mberr replied:
ccfeb08

cthoyt committed:
Update name for clarity

cthoyt committed:
Fix british

cthoyt committed:
Mark slow test

cthoyt committed:
Update .travis.yml

cthoyt committed:
Update .travis.yml

mberr commented on tests/test_interactions.py:
```diff
@@ -263,7 +263,7 @@ def test_scores(self):
             # calculate manually
             scores_f_manual = self._exp_score(**kwargs).view(-1)
-            assert torch.allclose(scores_f_manual, scores_f)
+            assert torch.allclose(scores_f_manual, scores_f), f'Diff: {scores_f_manual - scores_f}'
```
Locally, for me, using `py.test` shows the arguments that lead to failure one line above the assertion error,

mberr committed:
Extend append_weight_regularizer input types

mberr committed:
Re-introduce ConvKB regularizer comment, and use append_weight_regularizer

mberr committed:
Cleanup utils.__all__

mberr committed:
Add comment to test_scores

cthoyt committed:
Merge branch 'master' into add_interaction_function_2

cthoyt committed:
Fix bug in upgrading

cthoyt committed:
Bandage on tricky typing

cthoyt committed:
Make typing correct

It's possible that regularization could be turned off on ConvKB (though not correct) so this makes mypy happy

mberr commented on src/pykeen/models/base.py:
```diff
@@ -1176,13 +1176,13 @@ def append_weight_regularizer(
         # normalize input
         if isinstance(parameter, (str, nn.Parameter)):
-            parameter = list(parameter)
+            parameter = [parameter]
```
@cthoyt will this also work for `Iterable[nn.Parameter]` like returned by `module.parameters()`?

cthoyt replied:
if the parameter is a string and you run list() on it, you get something you weren't expecting.

```python
>>> parameter = 'hello'
>>> list(parameter)
['h', 'e', 'l', 'l', 'o']
```

It won't even work if `isinstance(parameter, nn.Parameter)` becuase nn.Parameter isn't implementing the iterable interface (I think). If it's an iterable of parameters, then this code won't get run because it won't pass the isinstance check

cthoyt committed:
Merge branch 'master' into add_interaction_function_2

mberr commented on src/pykeen/models/unimodal/conv_kb.py:
```diff
@@ -96,10 +96,12 @@ def __init__(
+            self.append_weight_regularizer(
+                parameter=self.interaction.parameters(),
+                regularizer=regularizer,
+            )
```
making mypy happy is sometimes dangerous, cf. https://github.com/pytorch/pytorch/pull/42356

cthoyt replied:
haha you were the one who wrote that? I've read that code before 😆

cthoyt replied:
But here I don't think that this is actually wrong. We do need to handle the possibility that the regularizer is None, since we're moving away from the NopRegularizer

mberr replied:
I didn't write the first part, only moved the method outside :smile:

mberr replied:
But it didn't solve all issues, cf. https://github.com/pytorch/pytorch/pull/42356#issuecomment-676865819

cthoyt committed:
Fix typo

mberr committed:
Fix unittest catching error

mberr committed:
Merge remote-tracking branch 'origin/add_interaction_function_2' into add_interaction_function_2

mberr committed:
Rename test case

mberr committed:
Test multiple different initializations for test_scores

mberr committed:
Fix typo

mberr committed:
Add TODO

mberr committed:
Remove wrong TODO

mberr committed:
Equivalentl rename dimensions for Tucker functional

mberr committed:
Fix exp_score for TuckerTests

mberr committed:
Fix some darglint issues for nn.emb docstrings

mberr committed:
Add missing parameter to docstring

mberr committed:
Improve docstring of _calculate_missing_shape_information

mberr committed:
Add raises to docstring

mberr committed:
Add raises to docstring

mberr committed:
Fix docstring for _add_dim

cthoyt committed:
asaf

cthoyt committed:
Improve logging on tests that sometimes fail

mberr committed:
Add reset_parameters to InteractionTests

mberr commented on tests/test_interactions.py:
```diff
@@ -181,7 +181,12 @@ def test_score_t_slicing(self):
+        self._check_close_scores(scores=scores, scores_no_slice=scores_no_slice)
+
+    def _check_close_scores(self, scores, scores_no_slice):
+        self.assertFalse(torch.isnan(scores).any(), msg=f'Normal scores had nan:\n\t{scores}')
```
@cthoyt there is also `torch.isfinite` additionally checking for `+/-inf`

cthoyt replied:
Nice, should we add that in addition to the nan check? Or do you mean instead?

mberr replied:
I think instead is fine

mberr commented on tests/test_interactions.py:
```diff
@@ -238,7 +243,9 @@ def test_forward(self):
                 if small_batch_size and has_batch_norm:
-                    logger.warning(f"Skipping test for shapes {hs}, {rs}, {ts}")
+                    logger.warning(
+                        f"Skipping test for shapes {hs}, {rs}, {ts} becuase too small batch size for batch norm",
```
typo

cthoyt replied:
you wouldn't believe how often I typo `because`

mberr committed:
Use isfinite to also check for +/-inf

mberr committed:
Fix typo

mberr committed:
Add shape to RepresentationModule

mberr committed:
Add max_id to RepresentationModule

lvermue commented on src/pykeen/pipeline.py:
```diff
@@ -709,6 +754,12 @@ def pipeline(  # noqa: C901
+    interaction_kwargs: Optional[Mapping[str, Any]] = None,
+    # Representations
+    entity_representations: OneOrMany[Union[None, RepresentationModule, EmbeddingSpecification]],
+    relation_representations: OneOrMany[Union[None, RepresentationModule, EmbeddingSpecification]],
```
@mberr 
Can you update the docstring accordingly?
Currently none of those examples would work.

mberr replied:
Yes, that is due to being only partially implemented. If we want to have the pipeline changes in this PR, we need to finish that. Otherwise, f30b73b goes into a new PR, and 
c1b6194 is the HEAD for this PR.

lvermue replied:
I vote for it being done in this PR, since those changes are directly related to the overall model API changes of this PR.

mberr replied:
I moved the pipeline changes to a new PR #163, which targets this branch. Thereby, we can keep them separated and decide which parts we want to have in master later without blocking progress.

mberr committed:
Simplify LCWA training loop

mberr committed:
Fix simplification

mberr committed:
Let training instance creation respect use_tqdm

mberr committed:
More explicit viewing

cthoyt committed:
Merge branch 'master' into add_interaction_function_2

cthoyt committed:
Add mypy testing

mberr committed:
Modify all interaction functions to use the broadcasted format

mberr committed:
Inline nop function

mberr committed:
Use matmul instead of einsum

mberr committed:
rename file

mberr committed:
Adjust get_in_canonical_shape

mberr committed:
Fix _get_representations

mberr committed:
Allow forwarding embedding specifications to ERModel

mberr committed:
Update DistMult

mberr committed:
Better integration of complex embeddings

mberr committed:
Update ConvKB

mberr committed:
Update ERMLP - KG2E

mberr committed:
Update NTN - RotatE

mberr committed:
Fix SimplE

mberr committed:
Update SE - UM

mberr committed:
Remove abstract base classes

mberr committed:
Fix DistMult and broken imports

mberr committed:
Fix broken annotations

mberr committed:
Fix _prepare_representation_module_list

mberr committed:
Fix _prepare_representation_module_list - again

mberr committed:
Fix SE shape

cthoyt committed:
Comment out unused code

@mberr why was this not used?

cthoyt committed:
pass flake8

mberr committed:
Fix ConvKB

mberr committed:
Merge remote-tracking branch 'origin/add_interaction_function_2' into add_interaction_function_2

mberr committed:
inline sizes for ConvE

mberr committed:
Fix ConvE init

mberr committed:
Fix ConvE revert reshaping

mberr committed:
Fix TransR functional form

cthoyt committed:
Add generics to LiteralModel

cthoyt committed:
Style cleanup

cthoyt committed:
Update handling fo embedding dimension in literal models

mberr committed:
Fix usage of chain_matmul

it only works for 2D matrices

mberr committed:
Fix NTN einsum eq

mberr committed:
Fix TransD

mberr committed:
Fix SimplEInteraction

mberr committed:
Fix RGCNRepresentations

mberr committed:
Merge remote-tracking branch 'origin/add_interaction_function_2' into add_interaction_function_2

mberr committed:
Add verification to representations utility

cthoyt committed:
Pass flake8 and mypy

mberr committed:
Merge remote-tracking branch 'origin/add_interaction_function_2' into add_interaction_function_2

cthoyt committed:
hsh

mberr committed:
Add shapes verification to passed representations

mberr committed:
Merge remote-tracking branch 'origin/add_interaction_function_2' into add_interaction_function_2

mberr committed:
Fix superclass order for ERModel

@cthoyt please check

mberr committed:
Improve error messages

mberr committed:
Extract canonical shape utility

cthoyt committed:
Pass flake8 / mypy

mberr committed:
Simplify score_* for Interaction

mberr committed:
Merge remote-tracking branch 'origin/add_interaction_function_2' into add_interaction_function_2

mberr committed:
Skip shape tests with non-consistent head entity/tail entity numbers

mberr committed:
fix TransR functional

mberr committed:
Adjust InteractionModule and tests

mberr committed:
Update strip dim utility

mberr committed:
Fix shapes for NTN

mberr committed:
Fix ProjE

mberr committed:
Fix HolE

mberr committed:
Fix HolE

mberr committed:
Move generic test classes

mberr committed:
Add shape information to SimplE module

mberr committed:
Add SimplE test

mberr committed:
Fix TransR relation projection matrices

mberr committed:
Fix SE

mberr committed:
For flake

mberr committed:
Do not use dictionary

mberr committed:
Add annotation for xx

mberr committed:
Give mypy a type for xx

cthoyt committed:
Update minimum python version

cthoyt committed:
Update representation.py

mberr committed:
Extend Embedding tests

mberr committed:
Fix Embedding.forward for >1d embeddings

mberr committed:
Add tests for LiteralRepresentations

mberr committed:
Add utility for expected_canonical_shape

mberr committed:
Improve documentation of tests

mberr committed:
Add test test for representations

mberr committed:
Add test for EmbeddingSpecification

mberr committed:
Add test for constructor error raising

mberr committed:
Add test for initializer

mberr committed:
Add documentation

mberr committed:
Add trailing comma

mberr committed:
Add tests for normalizer

mberr committed:
Add tests for constrainer; simplify code

mberr committed:
Extract method

mberr committed:
Delete unused variable

mberr committed:
Subclass LiteralRepresentationsTests from correct test super class

mberr committed:
Move RGCN representations to representation.py

mberr committed:
Add tests for RGCN representations

mberr committed:
Fix instantiation of RGCN test instance

mberr committed:
Add exception to RGCN representations for indices in wrong shape

mberr committed:
Inline constants

mberr committed:
Rename constant

mberr committed:
Add upgrade_to_sequence to export of utils.py

mberr committed:
Use complex dtype for ComplEx interaction

mberr committed:
use tensor product

mberr committed:
Extract common utility for elementwise tensor combinations

mberr committed:
Add elementwise combination optimization

mberr committed:
Add tests for broadcasted combination

mberr committed:
Add mark.slow

mberr committed:
Add two implementation variants of complex to functional

mberr committed:
Add another complex variant

mberr committed:
Add cost estimation to rotate forward

mberr committed:
Add draft of functional interaction function benchmark

mberr committed:
change benchmark script

mberr committed:
Fix variant creation

cthoyt committed:
Store results better

cthoyt committed:
Cleaner generation of stateless interaction class

cthoyt committed:
Update interaction benchmark

mberr committed:
Add more options

mberr committed:
Add gpu support to kernel benchmark

cthoyt committed:
Pass flake8

cthoyt committed:
Add docs stubs

cthoyt committed:
Pass mypy

cthoyt committed:
Merge branch 'master' into add_interaction_function_2

cthoyt committed:
Merge branch 'master' into add_interaction_function_2

mberr committed:
Merge remote-tracking branch 'origin/add_interaction_function_2' into add_interaction_function_2

mberr committed:
Delete unused variable

mberr committed:
extract common computation for KG2E similarities

mberr committed:
Pass h/t down to KG2E similarity functions

mberr committed:
Add docstring

mberr committed:
Add tensor_sum / tensor_product to __all__

mberr committed:
Do not use tensor_sum when it is clear that all summands are in the same shape

mberr committed:
Do not use tensor_sum when it is clear that all summands are in the same shape

mberr committed:
short-cut tensor_sum for less than 3 tensors

mberr committed:
Add benchmarking script for tensor_sum

mberr committed:
Remove unused method

mberr committed:
Update utils benchmark with all shape combinations of tensorsum

mberr committed:
Add click

mberr committed:
fixes

mberr committed:
Do not use tensor_sum for 2-element list

mberr committed:
Add caching test for get_optimal_sequence

mberr committed:
Fix test name

mberr committed:
Add additional heuristic for score_hrt case

mberr committed:
do not benchmark two-element tensor_sums

cthoyt committed:
Remove amo from model

cthoyt committed:
More manual amo merge

cthoyt committed:
Cleanup

cthoyt committed:
Merge remote-tracking branch 'origin/master' into add_interaction_function_2

# Conflicts:
#	src/pykeen/models/base.py
#	src/pykeen/models/multimodal/complex_literal.py
#	src/pykeen/models/multimodal/distmult_literal.py
#	src/pykeen/models/unimodal/complex.py
#	src/pykeen/models/unimodal/conv_e.py
#	src/pykeen/models/unimodal/distmult.py
#	src/pykeen/models/unimodal/ermlp.py
#	src/pykeen/models/unimodal/ermlpe.py
#	src/pykeen/models/unimodal/kg2e.py
#	src/pykeen/models/unimodal/ntn.py
#	src/pykeen/models/unimodal/proj_e.py
#	src/pykeen/models/unimodal/rescal.py
#	src/pykeen/models/unimodal/rgcn.py
#	src/pykeen/models/unimodal/simple.py
#	src/pykeen/models/unimodal/structured_embedding.py
#	src/pykeen/models/unimodal/trans_d.py
#	src/pykeen/models/unimodal/trans_e.py
#	src/pykeen/models/unimodal/trans_h.py
#	src/pykeen/models/unimodal/trans_r.py
#	src/pykeen/models/unimodal/tucker.py
#	src/pykeen/models/unimodal/unstructured_model.py
#	tests/test_early_stopping.py
#	tests/test_evaluators.py

cthoyt committed:
Pass lint

mberr committed:
Fix ER

mberr committed:
Adapt unittests for KL div similarity

mberr committed:
add test in test

mberr committed:
reformat

mberr committed:
reformat

mberr committed:
Move debug implementation to sim.py

mberr committed:
Fix bug in KL

mberr committed:
Remove unused variables

mberr committed:
Fix flake8

mberr committed:
Fix unittest

mberr committed:
Add skip_cls

mberr committed:
Fix unittest for RotatEInteraction

mberr committed:
circumvent accumulate(..., initial=x) since it is unavailable in Python 3.7

cthoyt committed:
More unittesting

hopefully this gives some better logging

cthoyt committed:
Fix mocking in test

Apparently this relied on an unstable API and the structure of the unittest.mock._Call class, which subclasses tuple. I switched it to using index based nomenclature instead

mberr commented on tests/test_nn.py:
```diff
@@ -150,15 +150,27 @@ def _check_call(
+        # Make sure tuple at least has positional arguments, could be 3 if kwargs available
+        self.assertLessEqual(2, call_size)
 
+        if call_size == 2:
```
@cthoyt add some comment why we need this? (and that we might get rid of it, once Python 3.7 support is dropped sometime in the future)

mberr committed:
Use autorange instead of fixed number of samples

mberr committed:
Merge remote-tracking branch 'origin/master' into add_interaction_function_2

mberr committed:
Merge remote-tracking branch 'origin/master' into add_interaction_function_2

# Conflicts:
#	MANIFEST.in
#	docs/source/extending/index.rst
#	src/pykeen/models/base.py
#	src/pykeen/models/multimodal/complex_literal.py
#	src/pykeen/models/multimodal/distmult_literal.py
#	src/pykeen/models/unimodal/complex.py
#	src/pykeen/models/unimodal/conv_e.py
#	src/pykeen/models/unimodal/conv_kb.py
#	src/pykeen/models/unimodal/distmult.py
#	src/pykeen/models/unimodal/ermlp.py
#	src/pykeen/models/unimodal/ermlpe.py
#	src/pykeen/models/unimodal/hole.py
#	src/pykeen/models/unimodal/kg2e.py
#	src/pykeen/models/unimodal/ntn.py
#	src/pykeen/models/unimodal/proj_e.py
#	src/pykeen/models/unimodal/rescal.py
#	src/pykeen/models/unimodal/rgcn.py
#	src/pykeen/models/unimodal/simple.py
#	src/pykeen/models/unimodal/structured_embedding.py
#	src/pykeen/models/unimodal/trans_d.py
#	src/pykeen/models/unimodal/trans_e.py
#	src/pykeen/models/unimodal/trans_h.py
#	src/pykeen/models/unimodal/trans_r.py
#	src/pykeen/models/unimodal/tucker.py
#	src/pykeen/models/unimodal/unstructured_model.py
#	src/pykeen/nn/emb.py
#	src/pykeen/triples/triples_factory.py
#	src/pykeen/triples/triples_numeric_literals_factory.py
#	src/pykeen/typing.py
#	src/pykeen/utils.py
#	tests/test_models.py
#	tests/test_pipeline.py
#	tests/test_utils.py
#	tox.ini

mberr committed:
Fix imports for triples_factory.py

mberr committed:
fix more imports

mberr committed:
Fix ComplEx constructor merge

mberr committed:
More post-merge fixes

mberr committed:
Extract literal interaction wrapper

mberr committed:
fix complex

cthoyt committed:
Merge branch 'master' into add_interaction_function_2

mberr committed:
Add work load generation for fast SLCWA

mberr committed:
Add docstring

mberr committed:
Move utility method

mberr committed:
Cleanup interaction benchmark script

mberr committed:
add another complex variant

mberr committed:
add fast flag

mberr committed:
add shuffle option for more reliable meta time estimates

mberr committed:
Also measure memory

mberr committed:
Remove viz

mberr committed:
fix variant

mberr committed:
Use single progress bar

mberr committed:
Extend docstring

mberr committed:
Add einsum variant

mberr committed:
Fix device error in KL divergence

mberr committed:
Do not use root logger

cthoyt committed:
Update tox.ini

cthoyt committed:
Pass flake8

cthoyt committed:
Update functional.py

mberr committed:
Fix tucker without explicit relation dim

mberr committed:
Merge remote-tracking branch 'origin/add_interaction_function_2' into add_interaction_function_2

cthoyt committed:
Merge branch 'master' into add_interaction_function_2

cthoyt committed:
Pass mypy

Trigger CI

cthoyt committed:
Bring back configurable regularization

Trigger CI

cthoyt committed:
Update pipeline.py

cthoyt committed:
Now I remember why I did this :laughing:

Gotta revert this

Trigger CI

mberr committed:
add another complex variant

mberr committed:
Merge remote-tracking branch 'origin/add_interaction_function_2' into add_interaction_function_2

cthoyt committed:
Update classvars

cthoyt committed:
More types

cthoyt committed:
More typing

mberr committed:
Add size information data class

mberr committed:
use utility for ER-MLP

mberr committed:
Update docstring

mberr committed:
Move variants to module

mberr committed:
extract common code

cthoyt commented on tests/test_interactions.py:
```diff
@@ -0,0 +1,618 @@
+
+    cls = pykeen.nn.modules.SimplEInteraction
+
+    def _exp_score(self, h, r, t, h_inv, r_inv, t_inv, clamp) -> torch.FloatTensor:
```
Is this right? I thought SimplE's interaction function was
```
0.5 * (distmult_interaction(h, r, t) + dismult_interaction(t_inv, r_inf, h_inv)
```

The h and t get switched for the inverse one

cthoyt committed:
Code cleanup

cthoyt committed:
Update test_models.py

cthoyt committed:
More cleanup

cthoyt committed:
Update losses.py

cthoyt committed:
Update typing.py

cthoyt committed:
Update losses.py

cthoyt committed:
Update tests.yml

cthoyt committed:
Update test_utils.py
